src/realm/sync/noinst/client_reset.cpp

```diff
@@ -40,169 +92,198 @@ bool _copy_list(Lst<T>& src, Lst<T>& dst)
-
-    bool updated = false;
-    size_t len_src = ll_src.size();
-    size_t len_dst = ll_dst.size();
-
-    size_t prefix_len, suffix_len;
 
-    for (prefix_len = 0; prefix_len < len_src && prefix_len < len_dst; ++prefix_len) {
-        auto ndx_src = ll_src.get(prefix_len);
-        auto ndx_dst = ll_dst.get(prefix_len);
-        auto ndx_converted = convert_ndx(ndx_src);
-        if (ndx_converted != ndx_dst)
-            break;
+    std::sort(to_delete.begin(), to_delete.end());
+    for (auto it = to_delete.rbegin(); it != to_delete.rend(); ++it) {
+        dst->erase_any(dst->get_any(*it));
     }
+    for (auto ndx : to_insert) {
+        Mixed converted_src = convert.src_to_dst(src->get_any(ndx));
+        dst->insert_any(converted_src);
+    }
+    return to_delete.size() || to_insert.size();
+}
 
-    for (suffix_len = 0; prefix_len + suffix_len < len_src && prefix_len + suffix_len < len_dst; ++suffix_len) {
-        auto ndx_src = ll_src.get(len_src - 1 - suffix_len);
-        auto ndx_dst = ll_dst.get(len_dst - 1 - suffix_len);
-        auto ndx_converted = convert_ndx(ndx_src);
-        if (ndx_converted != ndx_dst)
+bool copy_dictionary(const Obj& src_obj, Obj& dst_obj, InterRealmValueConverter& convert)
+{
+    Dictionary src = src_obj.get_dictionary(convert.source_col());
+    Dictionary dst = dst_obj.get_dictionary(convert.dest_col());
+
+    std::vector<size_t> sorted_src, sorted_dst, to_insert, to_delete;
+    constexpr bool ascending = true;
+    src.sort_keys(sorted_src, ascending);
+    dst.sort_keys(sorted_dst, ascending);
+
+    size_t dst_ndx = 0;
+    size_t src_ndx = 0;
+    while (src_ndx < sorted_src.size()) {
+        if (dst_ndx == sorted_dst.size()) {
+            // if we have reached the end of the dst items, all remaining
+            // src items should be added
+            while (src_ndx < sorted_src.size()) {
+                to_insert.push_back(sorted_src[src_ndx++]);
+            }
             break;
-    }
+        }
 
-    if (len_src > len_dst) {
-        // New elements must be inserted in ll_dst.
-        for (size_t i = prefix_len; i < prefix_len + (len_src - len_dst); ++i) {
-            auto ndx_src = ll_src.get(i);
-            auto ndx_converted = convert_ndx(ndx_src);
-            ll_dst.insert(i, ndx_converted);
-            updated = true;
+        auto src_val = src.get_pair(sorted_src[src_ndx]);
+        while (dst_ndx < sorted_dst.size()) {
+            auto dst_val = dst.get_pair(sorted_dst[dst_ndx]);
+            int cmp = src_val.first.compare(dst_val.first);
+            if (cmp == 0) {
+                Mixed converted_src = convert.src_to_dst(src_val.second);
+                cmp = converted_src.compare(dst_val.second);
+            }
+            if (cmp == 0) {
+                // equal: advance both src and dst
+                ++dst_ndx;
+                ++src_ndx;
+                break;
+            }
+            else if (cmp < 0) {
```

Comment:
Not sure I understand the logic here. You must differentiate between when keys differ and when value differ. Seems to be copy-pasted from Set.

Reply:
Yes, it is mostly the same as the logic for Set above. The difference being that the value of `cmp` is first assigned to the comparison of keys, and only if the keys are equal, it is assigned again to the comparison of values. See the use of `.first` (key) and `.second` (value) above. I think it is doing the correct thing?

Reply:
It might be that the end result in some mysterious way end up being right, but I think the operations that end up being executed are not what I would expect. If the destination contains {"adam", 10} and the source contains {"adam", 2} or {"adam", 20], I would in both cases expect an insert of the source value in the destination and not a delete followed by an insert. Please refer to branch je/tmp to see how I would implement it.

Reply:
Yes, good catch you are right. I will use your fix. Thanks!